[docs] Update router 101 to include protected routes (#37815)
expo/expo: docs/pages/router/basics/common-navigation-patterns.mdx

## Authenticated users only: protected routes

For mobile apps requiring authentication, you will likely have a set of routes that should only be accessible to authenticated users.

> Say we want to have a bottom tabs layout, a sign in page, a create account page, and a modal that should only be visible to authenticated users:

Reviewer: ```suggestion
For example, consider the following navigation tree in which you have a bottom tabs layout, a sign-in page, a create account page, and a modal that should only be visible to authenticated users:
```

Let's use the consistent pattern when describing an example here.

<FileTree
  files={[
    ['app/_layout.tsx', 'Root layout'],
    ['app/(tabs)/_layout.tsx'],
    [
      'app/(tabs)/index.tsx',
      <span>
        Protected <Lock01Icon className="mb-1 inline" />
      </span>,
    ],
    [
      'app/(tabs)/settings.tsx',
      <span>
        Protected <Lock01Icon className="mb-1 inline" />
      </span>,
    ],
    ['app/sign-in.tsx'],
    ['app/create-account.tsx'],
    [
      'app/modal.tsx',
      <span>
        Protected <Lock01Icon className="mb-1 inline" />
      </span>,
    ],
  ]}
/>

> When your app is first launched, router will try to open the root index, **app/(tabs)/index.tsx**. If we wrap this screen in a `Stack.Protected` with the `guard={false}`, the screen will become inaccessible and the next available screen will be opened instead. In our example, the `sign-in` screen will be opened, since it is the next available route.

Reviewer: ```suggestion
When your app is first launched, the router will try to open the root index, **app/(tabs)/index.tsx**. If you wrap this screen in a `Stack.Protected` with the `guard={false}`, the screen will become inaccessible and the next available screen will be opened instead. In this example, the `sign-in` screen will be opened, since it is the next available route.
```

Let's avoid using first-person voice as per the writing style guide: https://github.com/expo/expo/blob/main/guides/Expo%20Documentation%20Writing%20Style%20Guide.md#write-in-the-second-person

```tsx app/_layout.tsx
import { Stack } from 'expo-router';
import { useAuthState } from '@/utils/authState';

export default function RootLayout() {
  const { isLoggedIn } = useAuthState();

  return (
    <Stack>
      <Stack.Protected guard={isLoggedIn}>
        <Stack.Screen name="(tabs)" />
        <Stack.Screen name="modal" />
      </Stack.Protected>

      <Stack.Protected guard={!isLoggedIn}>
        <Stack.Screen name="sign-in" />
        <Stack.Screen name="create-account" />
      </Stack.Protected>
    </Stack>
  );
}
```

> This way you can fetch your auth state from a store and show the appropriate screens. If the auth state changes, the layout will re-render, so if `isLoggedIn` changes from `false` to `true`, the app will automatically navigate to the root of the `(tabs)` group.

Reviewer: ```suggestion
This way, you can fetch your auth state from a store and show the appropriate screens. If the auth state changes, the layout will re-render, so if `isLoggedIn` changes from `false` to `true`, the app will automatically navigate to the root of the `(tabs)` group.
```

> Another benefit of protected routes is that they are checked even if you deep link into a page directly. For example if an unautenticated user deep links into the modal screen above, they will be redirected to the sign-in page.

Reviewer: ```suggestion
Another benefit of protected routes is that they are checked even if you deep link into a page directly. For example, if an unauthenticated user deep links into the modal screen above, they will be redirected to the sign-in page.
```

Typo.
